doc/manpages/ccninitkeystore.1
'\" t
.\"     Title: ccninitkeystore
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: 01/10/2014
.\"    Manual: \ \&
.\"    Source: \ \& 0.8.2
.\"  Language: English
.\"
.TH "CCNINITKEYSTORE" "1" "01/10/2014" "\ \& 0\&.8\&.2" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ccninitkeystore \- Initialize a CCNx keystore with given parameters
.SH "SYNOPSIS"
.sp
\fBccninitkeystore\fR [\-h] [\-f] [\-u \fIusername\fR] [\-l \fIkeylength\fR] [\-v \fIvalidity\fR] [\fIdirectory\fR]
.SH "DESCRIPTION"
.sp
The \fBccninitkeystore\fR utility creates or overwrites the CCNx user keystore\&. The keystore is always written to the file \&.ccnx_keystore\&. If a directory is given the keystore will be written to the subdirectory \&.ccnx in that directory\&. Otherwise it will be written in $HOME/\&.ccnx\&. The \&.ccnx subdirectory will be created if it doesn\(cqt exist\&.
.SH "OPTIONS"
.PP
\fB\-h\fR
.RS 4
Print usage to stderr and exit\&.
.RE
.PP
\fB\-f\fR
.RS 4
Overwrite the existing keystore if one exists\&.
.RE
.PP
\fB\-u\fR \fIusername\fR
.RS 4
Username for this keystore\&. Defaults to username of effective uid\&.
.RE
.PP
\fB\-l\fR \fIkeylength\fR
.RS 4
Length of RSA key to be generated\&. Default is 1024 bits\&.
.RE
.PP
\fB\-v\fR \fIvalidity\fR
.RS 4
Nunber of days that the certificate should be valid\&. Default is 30\&.
.RE
.SH "EXIT STATUS"
.PP
\fB0\fR
.RS 4
Success
.RE
.PP
\fB1\fR
.RS 4
Failure (diagnostic error written on stderr)
.RE
.SH "SEE ALSO"
.sp
\fBccninitaeskeystore\fR(1)
.SH "AUTHOR"
.sp
Nick Briggs


NEWS
Changes between 0.8.1 and 0.8.2:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/21 for more details on CCNx 0.8.2.

- This release features a refactored strategy layer, making it straightforward to add new
  strategies in a modular way.  Alongside the traditional "default" strategy there are
  two new examples, a "parallel" strategy that sends interests on all available faces
  and a "loadsharing" strategy that sends each arriving interest to just one of the
  available faces.  Strategies may be selected by prefix.  This selection may be
  done using new subcommands to the ccndc utility.

	* Feature #100929: Implement Junit test suites to facilitate bulk testing.
	* Feature #100983: strategy layer refactoring
	* Feature #100984: strategy labels on prefixes
	* Feature #100990: Restore SYNC_NOISY to INFO
	* Feature #100992: Add test-security to standard java tests
	* Feature #100993: Add test-versioning back into standard java tests
	* Feature #100994: Remove unused cruft from build.xml
	* Feature #101036: java sync should use Deque operations instead of Stack as recommended in the javadocs
	* Feature #101039: UDP mode for java netmanagers is not tested in standard testing
	* Feature #101053: ccncatchunks2 could use improved hole-filling and interest lifetime management
	* Bug #101030: build.xml setup of ccnr environment is incorrect
	* Bug #101033: Handling of extremely large values of MaxSuffixComponents is incorrect
	* Bug #101037: ccn_seqwriter library code should not limit the block size to 4096 bytes
	* Bug #101042: CcnFileProxy uses deprecated methods
	* Bug #101046: VLC 2.1.0 release necessitates change in plugin Makefile
	* Bug #101048: Do not rely on accept(2) returned addr for recognizing a unix-domain socket
	* Bug #101052: Unbreak objtree build target
	* Bug #101057: ccn_sign_content / ccn_chk_signing_params do not properly incorporate a timestamp in a template via ccn_signing_params
	* Bug #101058: ccndc in daemon mode does not exit cleanly when ccnd goes away
	* Bug #101059: testTimeouts a little too strict
	* Bug #101060: ccnd cache lookups should take a controllable amount of work
	* Bug #101061: ccnd minimum interest lifetime problem
	* Bug #101062: some ccnd logging missing check for logging enabled before calling expensive log procedure
	* Support #101047: Prepare for ccnx-0.8.2 release

Changes between 0.8.0 and 0.8.1:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/20 for more details on CCNx 0.8.1.

- This release fixes a number of bugs, including a serious memory leak in ccnd.

	* Bug #101029: unbounded test log file growth
	* Bug #101032: ccnr should start at specified segment on checked start write
	* Bug #101034: No obvious documentation for how to start ccnr exists on website
	* Bug #101035: small leak in ccn_fetch should a realloc() fail
	* Bug #101038: Inconsistent encoding of zero-length name components
	* Bug #101043: ccnr r_store_lookup_backwards can fail
	* Bug #101044: valgrind reports memory leak in 0.8.0 ccnd
	* Bug #101049: ccnd can remove newly arrived content too quickly
	* Support #101045 Prepare for ccnx-0.8.1 release

Changes between 0.7.2 and 0.8.0:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/17 for more details on CCNx 0.8.0.

- The use of symmetric key HMACs in place of public/private key signatures
  is now an option in both libraries.  In situations where it can be used,
  it provides a much higher performance alternative.
- The content store in ccnd has been improved.  It now uses a flatname
  representation for indexing, in the same way that ccnr does.  Also, the
  underlying skiplist implementation has moved into the library, so that
  applications may take advantage of it in the future.  In the process,
  some old ccnd bugs relating to the handling of extreme values of
  FreshnessSeconds were fixed.
- The more readable escaping convention in URIs is now the default.
  If for whatever reason you prefer to use only percent-escaping in URIs, use

	CFLAGS='-g -DCCN_URI_DEFAULT_ESCAPE=4' ./configure

  before typing make.  Note that both forms of escaping are accepted as input.
- An example rc startup script is now included, but needs to be installed by
  hand.  It is coded for Ubuntu Linux, but should be easily adaptable to other
  systems.
- This release drops support for Java JDK 1.5.

  * Issues Resolved:
	* Bug #100118: ccnd does not handle FreshnessSeconds=0 correctly.
	* Bug #100119: ccnd should handle large values of FreshnessSeconds
	* Bug #100558: test-security does not pass
	* Bug #100897: timeout values  should be consistently long throughout java code 
	* Bug #100966: Incorrect processing of CCNS_SYNC_SCOPE=3
	* Bug #100977: Problem with verify in CCNInputStream when trying to retrieve key
	* Bug #100987: Non-portable use of local keyword in Bourne shell scripts
	* Bug #100998: VLC 2.1.x block_New definition removed
	* Bug #100999: java BLOCK_AFTER_FIRST_SEGMENT doesn't work in all cases
	* Bug #101000: sync_diff code should reset differencing data when aborting or completing a comparison
	* Bug #101004: QueuedContentHandler has a sync error
	* Bug #101005: CCNx Android Services build out of sync with API changes to CCNx C-lib
	* Bug #101007: ccn_resolve_version() should distinguish between no response and finding an unversioned object.
	* Bug #101010: ccnseqwriter command busy-loops when its ccnd goes away
	* Bug #101011: Compile warning when building new symmetric key code with openssl 0.9.8
	* Bug #101012: Improve recovery from node fetch timeout in java sync
	* Bug #101013: node fetch timeout in java sync implementation should be longer and settable

csrc/cmd/ccninitkeystore.c
/**
 * @file ccninitkeystore.c
 * Initialize a CCNx keystore with given parameters.
 *
 * A CCNx command-line utility.
 *
 * Copyright (C) 2011 Palo Alto Research Center, Inc.
 *
 * This work is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License version 2 as published by the
 * Free Software Foundation.
 * This work is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
 * for more details. You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <pwd.h>
#include <sys/stat.h>
#include <ccn/ccn.h>
#include <ccn/charbuf.h>
#include <ccn/keystore.h>

#define CCN_KEYSTORE_PASS "Th1s1sn0t8g00dp8ssw0rd."

static void
usage(const char *progname)
{
    fprintf(stderr,
            "%s [-h] [-f] [-u username] [-l keylength] [-v validity] [directory]\n"
            "   Initialize a CCNx keystore with given parameters\n", progname);
    fprintf(stderr,
            "   -h  Display this help message.\n"
            "   -f  Force overwriting an existing keystore. Default no overwrite permitted.\n" 
            "   -u username  Username for this keystore.  Default username of effective uid.\n"
            "   -l keylength  Length of RSA key to be generated.  Default 1024 bits.\n"
            "   -v validity  Number of days that certificate should be valid.  Default 30.\n"
            "   directory  Directory in which to create .ccnx/.ccnx_keystore. Default $HOME.\n"
            );
}

int
main(int argc, char **argv)
{
    int res;
    int opt;
    struct stat statbuf;
    char *dir;
    struct ccn_charbuf *keystore = NULL;
    int force = 0;
    char *user = NULL;
    char useruid[32];
    struct passwd *pwd = NULL;
    int keylength = 0;
    int validity = 0;
    
    while ((opt = getopt(argc, argv, "hfu:p:l:v:")) != -1) {
        switch (opt) {
            case 'f':
                force = 1;
                break;
            case 'u':
                user = optarg;
                break;
            case 'l':
                keylength = atoi(optarg);
                if (keylength < 512) {
                    fprintf(stderr, "%d: Key length too short for signing CCNx objects.\n", keylength);
                    exit(1);
                }
                break;
            case 'v':
                validity = atoi(optarg);
                if (validity < 0) {
                    fprintf(stderr, "%d: Certificate validity must be > 0.\n", validity);
                    exit(1);
                }
                break;
            case 'h':
            default:
                usage(argv[0]);
                exit(1);
        }
    }
    dir = argv[optind];
    if (dir == NULL){
        dir = getenv("HOME");
    }
    res = stat(dir, &statbuf);
    if (res == -1) {
        perror(dir);
        exit(1);
    } else if ((statbuf.st_mode & S_IFMT) != S_IFDIR) {
        errno = ENOTDIR;
        perror(dir);
        exit(1);
    }
    keystore = ccn_charbuf_create();
    ccn_charbuf_putf(keystore, "%s/.ccnx", dir);
    res = stat(ccn_charbuf_as_string(keystore), &statbuf);
    if (res == -1) {
        res = mkdir(ccn_charbuf_as_string(keystore), 0700);
        if (res != 0) {
            perror(ccn_charbuf_as_string(keystore));
            exit(1);
        }
    }
    ccn_charbuf_append_string(keystore, "/.ccnx_keystore");
    res = stat(ccn_charbuf_as_string(keystore), &statbuf);
    if (res != -1 && !force) {
        errno=EEXIST;
        perror(ccn_charbuf_as_string(keystore));
        exit(1);
    }
    if (user == NULL) {
        errno = 0;
        pwd = getpwuid(geteuid());
        if (pwd != NULL)
            user = pwd->pw_name;
        else {
            if (errno != 0) {
                perror("getpwuid");
                exit(1);
            }
            snprintf(useruid, sizeof(useruid), "uid%d", geteuid());
            user = useruid;
        }
    }
    res = ccn_keystore_file_init(ccn_charbuf_as_string(keystore), CCN_KEYSTORE_PASS,
                                 user, keylength, validity);
    if (res != 0) {
        if (errno != 0)
            perror(ccn_charbuf_as_string(keystore));
        else
            fprintf(stderr, "ccn_keystore_file_init: invalid argument\n");
        exit(1);
    }
    return(0);
}
